sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/ExpansionService.java

```diff
@@ -486,6 +504,11 @@ default InputT createInput(Pipeline p, Map<String, PCollection<?>> inputs) {
         .build();
   }
 
+  protected Pipeline createPipeline() {
+    pipelineOptions.setRunner(NoOpRunner.class);
```

Comment:
How or where does this fail if you just do "java -jar beam-sdks-java-io-expansion-service-<version>-SNAPSHOT.jar <port> --defaultEnvironmentType=PROCESS" without specifying a Runner here ?

Reply:
Runner is `@Validation.Required`, so it fails in `PipelineOptionsValidator#validate`.

Reply:
Exactly `PipelineOptions.java:301`

Reply:
How about just adding both "beam-sdks-java-io-expansion-service" and beam-runners-direct-java" to the classpath when starting up the expansion service ? No need to update dependencies of Beam. Following seems to work for me.

export CLASSPATH=beam-sdks-java-io-expansion-service-2.32.0-SNAPSHOT.jar:beam-runners-direct-java-2.32.0-SNAPSHOT.jar
java org.apache.beam.sdk.expansion.service.ExpansionService 12345

Reply:
The problem is that `:sdks:java:expansion-service` would (in fact) depend on :runners:direct-java`. It would require this at least for `testRuntimeOnly`, but actually all uses of this module would need to add that (or any other runner and specify --runner, but that will get ignored). I think that the biggest problem is that runner is marked `@Validation.Required`, athough it is not required (apparently, expansion service works without runner). That is the root issue and everything else seems like a (smaller or bigger) hack. Implementing a `NoOpRunner` seems like the smallest issue, as that does not leak or break users anyhow.

Reply:
Hmm, maybe just deleting the `@Validation.Required` from `runner` in `PipelineOptions` might be enough? What is the reasoning for having both `@Validation.Required` and `@Default.InstanceFactory` on the same field? That means that the field has default value - though will be present - and if it is left to the default, it will throw the exception as soon as the Pipeline will try to use it, right?

Reply:
Understand that. I think that the NoOpRunner (or if we rename it to ExpansionRunner) is actually better than DirectRunner, as mentioned in comment below.

Reply:
Hm, it depends on what you mean by 'works'. Yes, it will expand, but the expansion service really is not supposed to call the `run` method. That would definitely result in wrong behavior, even with DirectRunner.